File: Java/1647. Minimum Deletions to Make Character Frequencies Unique.java
class Solution {
    public int minDeletions(String s) {
        int[] counts = new int[26];
        for (char c : s.toCharArray()) {
            counts[c - 'a'] += 1;
        }
        Arrays.sort(counts);
        int ans = 0;
        for (int i = 25; i > 0; i--) {
            if (counts[i] <= counts[i - 1] && counts[i - 1] > 0) {
                int diff = counts[i] == 0 ? counts[i - 1] : counts[i - 1] - counts[i] + 1;
                counts[i - 1] -= diff;
                ans += diff;
            }
        }
        return ans;
    }
}

File: 2023/1647. Minimum Deletions to Make Character Frequencies Unique.cpp
class Solution {
public:
    int minDeletions(string s) {
        vector<int> counts(26);
        for (char c : s) {
            counts[c - 'a']++;
        }
        unordered_set<int> nums;
        int total = 0;
        for (int c : counts) {
            if (c == 0) {
                continue;
            }
            int n = c, diff = 0;
            while (nums.find(c - diff) != nums.end() && c - diff > 0) {
                diff++;
            }
            total += diff;
            if (c - diff != 0) {
                nums.insert(c - diff);
            }
        }
        return total;
    }
};